### PR TITLE
FastCell Example Fixes, Generalized trainer for both batch_first args

### PR DESCRIPTION
**Pull request by SachinG007**
Checked the changes with Aditya

### CODE REVIEW

File: pytorch/edgeml_pytorch/trainer/fastTrainer.py [hunk 1]
```diff
@@ -77,9 +77,13 @@ def computeLogits(self, input):
             logits = self.classifier(feats[-1, :])
         else:
             feats = self.RNN(input)
-            logits = self.classifier(feats[-1, :])
-
-        return logits, feats[:, -1]
+            if self.batch_first:
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/trainer/fastTrainer.py:80]: This doesn't handle LSTM.

Comment (SachinG007) [reply pytorch/edgeml_pytorch/trainer/fastTrainer.py]: Corrected accordingly

**Comment by SachinG007** [comment]
@oindrilasaha would be good if you can have a look at this PR

**Comment by adityakusupati** [comment]
@SachinG007 please incorporate the changes from #173 for the fixing of optimizer for FC. Earlier gradient updates were not happening for FC due to some convention mismatch. Please look at this: https://github.com/microsoft/EdgeML/pull/173/files#diff-7b39dde7dda6360cbf530db88f5b9f8dR12-R62 and incorporate it. 

Also, try to keep PRs for different things separate. When we are fixing fastcell_Example.py related stuff, let us stick to that. Thanks

**Comment by SachinG007** [comment]
@adityakusupati , incorporated the changes from PR#173.

File: examples/pytorch/FastCells/fastcell_example.py [hunk 2]
```diff
@@ -43,10 +43,17 @@ def main():
...
         "Timesteps have to be integer"
 
+    timeSteps = int(dataDimension / inputDims)
+    Xtrain = Xtrain.reshape((-1, timeSteps, inputDims))
+    Xtest = Xtest.reshape((-1, timeSteps, inputDims))
+
+    if not batch_first:
+        Xtrain = np.swapaxes(Xtrain, 0, 1)
+        Xtest = np.swapaxes(Xtest, 0, 1)
+
```
Comment (adityakusupati) [inline examples/pytorch/FastCells/fastcell_example.py:56]: This looks good.

File: pytorch/edgeml_pytorch/trainer/fastTrainer.py [hunk 3]
```diff
@@ -9,6 +9,14 @@
...
 import numpy as np
 
+class SimpleFC(nn.Module):
+    def __init__(self, input_size, num_classes, name="SimpleFC"):
+        super(SimpleFC, self).__init__()
+        self.FC = nn.Parameter(torch.randn([input_size, num_classes]))
+        self.FCbias = nn.Parameter(torch.randn([num_classes]))
+
+    def forward(self, input):
+        return torch.matmul(input, self.FC) + self.FCbias
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/trainer/fastTrainer.py:19]: I am not sure if we should place this here or make a different file or place in rnn.py. Ideally, this is the same as Bonsai.py or ProtoNN.py. We need to make a decision about placing this at the right place.

File: pytorch/edgeml_pytorch/trainer/fastTrainer.py [hunk 4]
```diff
@@ -50,43 +58,41 @@ def __init__(self, FastObj, numClasses, sW=1.0, sU=1.0,
...
-        self.optimizer = self.optimizer()
-
         self.RNN = BaseRNN(self.FastObj, batch_first=self.batch_first).to(self.device)
-
-        self.FC = nn.Parameter(torch.randn(
-            [self.FastObj.output_size, self.numClasses])).to(self.device)
-        self.FCbias = nn.Parameter(torch.randn(
-            [self.numClasses])).to(self.device)
-
+        self.simpleFC = SimpleFC(self.FastObj.output_size, self.numClasses).to(self.device)
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/trainer/fastTrainer.py:62]: a better name for the instance instead of self.simpleFC (so that we are generic enough to fit the classifier() function.

File: pytorch/edgeml_pytorch/trainer/fastTrainer.py [hunk 5]
```diff
@@ -50,43 +58,41 @@ def __init__(self, FastObj, numClasses, sW=1.0, sU=1.0,
...
         self.FastParams = self.FastObj.getVars()
+        self.optimizer = self.optimizer()
 
     def classifier(self, feats):
         '''
         Can be raplaced by any classifier
         TODO: Make this a separate class if needed
         '''
-        return torch.matmul(feats, self.FC) + self.FCbias
+        return self.simpleFC(feats)
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/trainer/fastTrainer.py:71]: Look at the above comment.

File: pytorch/edgeml_pytorch/trainer/fastTrainer.py [hunk 6]
```diff
@@ -50,43 +58,41 @@ def __init__(self, FastObj, numClasses, sW=1.0, sU=1.0,
...
             feats = self.RNN(input)
-            logits = self.classifier(feats[-1, :])
 
-        return logits, feats[:, -1]
+        if self.batch_first:
+                logits = self.classifier(feats[:, -1])
+                return logits, feats[:, -1]
+        else:
+                logits = self.classifier(feats[-1, :])
+                return logits, feats[-1, :]
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/trainer/fastTrainer.py:87]: This looks good.

File: pytorch/edgeml_pytorch/trainer/fastTrainer.py [hunk 7]
```diff
@@ -50,43 +58,41 @@ def __init__(self, FastObj, numClasses, sW=1.0, sU=1.0,
...
 
     def optimizer(self):
         '''
         Optimizer for FastObj Params
         '''
+        paramList = list(self.FastObj.parameters()) + list(self.simpleFC.parameters())
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/trainer/fastTrainer.py:93]: Check the characters in the line and break it using PEP8 if it is overshooting the guidelines.

File: pytorch/edgeml_pytorch/trainer/fastTrainer.py [hunk 8]
```diff
@@ -351,7 +357,13 @@ def train(self, batchSize, totalEpochs, Xtrain, Xtest, Ytrain, Ytest,
...
         fileName = str(self.FastObj.cellType) + 'Results_pytorch.txt'
         resultFile = open(os.path.join(dataDir, fileName), 'a+')
-        numIters = int(np.ceil(float(Xtrain.shape[0]) / float(batchSize)))
+        if self.batch_first:
+                self.timeSteps = Xtrain.shape[1]
+                self.numPoints = Xtrain.shape[0]
+        else:
+                self.timeSteps = Xtrain.shape[0]
+                self.numPoints = Xtrain.shape[1]
+        numIters = int(np.ceil(float(self.numPoints) / float(batchSize)))
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/trainer/fastTrainer.py:366]: This looks good.

File: pytorch/edgeml_pytorch/trainer/fastTrainer.py [hunk 9]
```diff
@@ -341,8 +347,8 @@ def saveParams(self, currDir):
             np.save(os.path.join(currDir, "Bo.npy"),
                     self.FastParams[self.totalMatrices + 3].data.cpu())
 
-        np.save(os.path.join(currDir, "FC.npy"), self.FC.data.cpu())
-        np.save(os.path.join(currDir, "FCbias.npy"), self.FCbias.data.cpu())
+        np.save(os.path.join(currDir, "FC.npy"), self.simpleFC.FC.data.cpu())
+        np.save(os.path.join(currDir, "FCbias.npy"), self.simpleFC.FCbias.data.cpu())
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/trainer/fastTrainer.py:351]: If we change self.simpleFC, please remember to change these.

File: pytorch/edgeml_pytorch/trainer/fastTrainer.py [hunk 10]
```diff
@@ -389,9 +396,12 @@ def train(self, batchSize, totalEpochs, Xtrain, Xtest, Ytrain, Ytest,
                           (header, msg, header), file=self.outFile)
 
                 k = shuffled[j * batchSize:(j + 1) * batchSize]
-                batchX = Xtrain[:, k, :]
+                if self.batch_first:
+                        batchX = Xtrain[k, :, :]
+                else:
+                        batchX = Xtrain[:, k, :]
+
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/trainer/fastTrainer.py:403]: This looks good.

File: pytorch/edgeml_pytorch/graph/rnn.py [hunk 11]
```diff
@@ -261,7 +261,7 @@ def __init__(self, input_size, hidden_size, gate_nonlinearity="sigmoid",
         self.zeta = nn.Parameter(self._zetaInit * torch.ones([1, 1]))
         self.nu = nn.Parameter(self._nuInit * torch.ones([1, 1]))
 
-        self.copy_previous_UW()
+        # self.copy_previous_UW()
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/graph/rnn.py:264]: Why is this commented out?  This supports sparsity for KWS codes. The fastcell_Example.py doesn't depend on it because this is a bit broken given the new pytorch updates. Check with Harsha.

File: pytorch/edgeml_pytorch/graph/rnn.py [hunk 12]
```diff
@@ -967,78 +967,142 @@ class BaseRNN(nn.Module):
     [batchSize, timeSteps, inputDims]
     '''
 
-    def __init__(self, cell: RNNCell, batch_first=False):
+    def __init__(self, cell: RNNCell, batch_first=False, cell_reverse: RNNCell=None, bidirectional=False):
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/graph/rnn.py:970]: What is this RNNCell=None? I don't know what it does. Why should there be an  RNNCell for this argument?

File: pytorch/edgeml_pytorch/graph/rnn.py [hunk 13]
```diff
@@ -967,78 +967,142 @@ class BaseRNN(nn.Module):
...
         super(BaseRNN, self).__init__()
-        self._RNNCell = cell
+        self.RNNCell = cell
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/graph/rnn.py:972]: Why did we flip self._RNNCell to self.RNNCell? Isn't it supposed to be private?

File: pytorch/edgeml_pytorch/graph/rnn.py [hunk 14]
```diff
@@ -967,78 +967,142 @@ class BaseRNN(nn.Module):
...
+    def __init__(self, cell: RNNCell, batch_first=False, cell_reverse: RNNCell=None, bidirectional=False):
         super(BaseRNN, self).__init__()
-        self._RNNCell = cell
+        self.RNNCell = cell 
         self._batch_first = batch_first
+        self._bidirectional = bidirectional
+        if cell_reverse is not None:
+            self.RNNCell_reverse = cell_reverse
+        elif self._bidirectional:
+            self.RNNCell_reverse = cell
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/graph/rnn.py:978]: Try to use if cell_reverse is None: first and then else.

```
if cell_reverse is None:
    self.RNNCell_reverse = cell
elif self._bidirectional:
    self.RNNCell_reverse = cell_reverse
```

File: pytorch/edgeml_pytorch/graph/rnn.py [hunk 15]
```diff
@@ -967,78 +967,142 @@ class BaseRNN(nn.Module):
...
 
     def getVars(self):
-        return self._RNNCell.getVars()
+        return self.RNNCell.getVars()
 
     def forward(self, input, hiddenState=None,
                 cellState=None):
         self.device = input.device
+        self.num_directions = 2 if self._bidirectional else 1
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/graph/rnn.py:986]: expand this conditional for easy readability.

File: pytorch/edgeml_pytorch/graph/rnn.py [hunk 16]
```diff
@@ -967,78 +967,142 @@ class BaseRNN(nn.Module):
...
+        # hidden
+        # for i in range(num_directions):
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/graph/rnn.py:988]: What are these comments?

File: pytorch/edgeml_pytorch/graph/rnn.py [hunk 17]
```diff
@@ -967,78 +967,142 @@ class BaseRNN(nn.Module):
...
         hiddenStates = torch.zeros(
                 [input.shape[0], input.shape[1],
-                 self._RNNCell.output_size]).to(self.device)
+                 self.RNNCell.output_size]).to(self.device)
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/graph/rnn.py:991]: All the lines after this are about bi-directional stuff. I would defer to @oindrilasaha to check all this stuff. She has to sign off on it as she uses them the most.

File: pytorch/edgeml_pytorch/graph/rnn.py (hunk 13 above)
Comment (SachinG007) [reply pytorch/edgeml_pytorch/graph/rnn.py:972]: I had similar doubt, but rnnpool flipped it so I too flipped for consistency. I can revert back.

File: pytorch/edgeml_pytorch/graph/rnn.py (hunk 14 above)
Comment (SachinG007) [reply pytorch/edgeml_pytorch/graph/rnn.py:978]: @oindrilasaha please have a look

File: pytorch/edgeml_pytorch/graph/rnn.py (hunk 13 above)
Comment (adityakusupati) [reply pytorch/edgeml_pytorch/graph/rnn.py:972]: Ask @oindrilasaha what is the reason she made it public instead of private. We already pass in the cell, so we have access to it if needed.

File: pytorch/edgeml_pytorch/trainer/fastTrainer.py [hunk 18]
```diff
@@ -9,6 +9,14 @@
 from edgeml_pytorch.graph.rnn import *
 import numpy as np
 
+class SimpleFC(nn.Module):
+    def __init__(self, input_size, num_classes, name="SimpleFC"):
+        super(SimpleFC, self).__init__()
+        self.FC = nn.Parameter(torch.randn([input_size, num_classes]))
+        self.FCbias = nn.Parameter(torch.randn([num_classes]))
```
Comment (adityakusupati) [inline pytorch/edgeml_pytorch/trainer/fastTrainer.py:16]: @SachinG007  change this to self.weight and self.bias.

File: pytorch/edgeml_pytorch/graph/rnn.py (hunk 11 above)
Comment (oindrilasaha) [reply pytorch/edgeml_pytorch/graph/rnn.py:264]: This line is causing segmentation fault in the rnnpool codes

File: pytorch/edgeml_pytorch/graph/rnn.py (hunk 12 above)
Comment (oindrilasaha) [reply pytorch/edgeml_pytorch/graph/rnn.py:970]: If the bidirectional pass doesn't share weights this passes an extra initialized RNNCell for doing the backward pass.

File: pytorch/edgeml_pytorch/graph/rnn.py (hunk 14 above)
Comment (oindrilasaha) [reply pytorch/edgeml_pytorch/graph/rnn.py:978]: the change suggested is also fine

File: pytorch/edgeml_pytorch/graph/rnn.py (hunk 13 above)
Comment (oindrilasaha) [reply pytorch/edgeml_pytorch/graph/rnn.py:972]: please make it as self._RNNCell everywhere, this change came since when i had pulled it was self.RNNCell.

File: pytorch/edgeml_pytorch/graph/rnn.py (hunk 11 above)
Comment (adityakusupati) [reply pytorch/edgeml_pytorch/graph/rnn.py:264]: Yeah! I know the code is broken, it needs to be fixed. @harsha-simhadri  used this in one of the codes so, we might have to fix it.

File: pytorch/edgeml_pytorch/graph/rnn.py (hunk 13 above)
Comment (SachinG007) [reply pytorch/edgeml_pytorch/graph/rnn.py:972]: Done

File: pytorch/edgeml_pytorch/graph/rnn.py (hunk 14 above)
Comment (SachinG007) [reply pytorch/edgeml_pytorch/graph/rnn.py:978]: Done, but not sure about logical equivalence. 
@oindrilasaha can you confirm once , if the suggestion is logically equivalent to the original

Comment (adityakusupati) [reply pytorch/edgeml_pytorch/graph/rnn.py:978]: My bad, I read it wrong. i thought it was an else and not life. revert this change

Comment (SachinG007) [reply pytorch/edgeml_pytorch/graph/rnn.py:978]: Keeping it the original one only finally.

File: pytorch/edgeml_pytorch/graph/rnn.py (hunk 12 above)
Comment (SachinG007) [reply pytorch/edgeml_pytorch/graph/rnn.py:970]: Should I resolve this conversation ? @oindrilasaha

File: pytorch/edgeml_pytorch/graph/rnn.py (hunk 17 above)
Comment (SachinG007) [reply pytorch/edgeml_pytorch/graph/rnn.py:991]: @oindrilasaha  any comments for the later section of the code ?

**Comment by adityakusupati** [comment]
@SachinG007 any updates on this PR?
